File: software/tests/test_wiegand_reader_getter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

HOST = "localhost"
PORT = 4223
UID = "HXg"

import time
from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_industrial_digital_in_4_v2 import BrickletIndustrialDigitalIn4V2

def interpret_wiegand26_data(data):
    if len(data) != 26:
        print('Bad Length {0}'.format(len(data)))
        return

    pe = 0
    for i in range(0, 13):
        pe = pe ^ data[i]

    po = 0
    for i in range(13, 26):
        po = po ^ data[i]

    if pe != 0 or po != 1:
        print('Paritiy Error')
        return

    fc = 0
    for i in range(1, 9):
        fc = fc << 1
        fc = fc | data[i]

    cc = 0
    for i in range(9, 25):
        cc = cc << 1
        cc = cc | data[i]

    print('Facility Code {0}, Card Code {1}'.format(fc, cc))

def cb_wiegand_error_count(framing_error_count, overflow_error_count):
    print('Framing Error Count {0}, Overflow Error Count {1}'.format(framing_error_count, overflow_error_count))

if __name__ == "__main__":
    ipcon = IPConnection() # Create IP connection
    idi = BrickletIndustrialDigitalIn4V2(UID, ipcon) # Create device object

    ipcon.connect(HOST, PORT) # Connect to brickd
    # Don't use device before ipcon is connected

    idi.register_callback(idi.CALLBACK_WIEGAND_ERROR_COUNT, cb_wiegand_error_count)

    idi.set_wiegand_reader_config(True, 26, 50) # 26 bit, with 50 ms timeout
    idi.set_wiegand_callback_config(False, False, True)

    while True:
        time.sleep(0.1)

        data = idi.read_wiegand_data()

        if len(data) > 0:
            interpret_wiegand26_data(data)

    ipcon.disconnect()

File: software/tests/test_wiegand_reader_getter.c
#define IPCON_EXPOSE_MILLISLEEP

#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_industrial_digital_in_4_v2.h"

#define HOST "localhost"
#define PORT 4223
#define UID "HXg"

void interpret_wiegand26_data(bool *data, uint16_t data_length) {
    if (data_length != 26) {
        printf("Bad length: %u\n", data_length);
        return;
    }

    uint8_t pe = 0;
    for (int i = 0; i < 13; ++i) {
        pe = pe ^ (data[i] ? 1 : 0);
    }

    uint8_t po = 0;
    for (int i = 13; i < 26; ++i) {
        po = po ^ (data[i] ? 1 : 0);
    }

    if (pe != 0 || po != 1) {
        printf("Paritiy Error\n");
        return;
    }

    uint8_t fc = 0;
    for (int i = 1; i < 9; ++i) {
        fc = fc << 1;
        fc = fc | (data[i] ? 1 : 0);
    }

    uint16_t cc = 0;
    for (int i = 9; i < 25; ++i) {
        cc = cc << 1;
        cc = cc | (data[i] ? 1 : 0);
    }

    printf("Facility Code %u, Card Number %u\n", fc, cc);
}

void cb_wiegand_error_count(uint32_t framing_error_count, uint32_t overflow_error_count, void *user_data) {
    (void)user_data; // avoid unused parameter warning

    printf("Framing Error Count %u, Overflow Error Count %u\n", framing_error_count, overflow_error_count);
}

int main(void) {
    // Create IP connection
    IPConnection ipcon;
    ipcon_create(&ipcon);

    // Create device object
    IndustrialDigitalIn4V2 idi4;
    industrial_digital_in_4_v2_create(&idi4, UID, &ipcon);

    // Connect to brickd
    if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
        fprintf(stderr, "Could not connect\n");
        return 1;
    }
    // Don't use device before ipcon is connected

    industrial_digital_in_4_v2_register_callback(&idi4,
                                                 INDUSTRIAL_DIGITAL_IN_4_V2_CALLBACK_WIEGAND_ERROR_COUNT,
                                                 (void (*)(void))cb_wiegand_error_count,
                                                 NULL);

    industrial_digital_in_4_v2_set_wiegand_reader_config(&idi4, true, 26, 20); // 26 bit, with 50 ms timeout
    industrial_digital_in_4_v2_set_wiegand_callback_config(&idi4, false, false, true);

    while (true) {
        millisleep(100);

        bool data[26];
        uint16_t data_length;
        industrial_digital_in_4_v2_read_wiegand_data(&idi4, data, &data_length);

        if (data_length > 0) {
            interpret_wiegand26_data(data, data_length);
        }
    }

    printf("Press key to exit\n");
    getchar();
    industrial_digital_in_4_v2_destroy(&idi4);
    ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
    return 0;
}
